astropy/units/tests/test_spectral_quantity.py

```diff
@@ -0,0 +1,166 @@
+
+    def test_doppler_set_parameters(self):
+
+        sq1 = SpectralQuantity(1 * u.km / u.s)
+
+        with pytest.raises(ValueError, match="doppler_convention should be one of optical/radio/relativistic"):
+            sq1.doppler_convention = 'banana'
+
+        assert sq1.doppler_convention is None
+
+        sq1.doppler_convention = 'radio'
+
+        assert sq1.doppler_convention == 'radio'
+
+        with pytest.raises(ValueError, match="doppler_convention has already been set, and cannot be changed"):
+            sq1.doppler_convention = 'optical'
+
+        assert sq1.doppler_convention == 'radio'
+
+        with pytest.raises(u.UnitsError, match="Argument 'value' to function 'doppler_rest' must be in units"):
+            sq1.doppler_rest = 5 * u.kg
+
+        sq1.doppler_rest = 5 * u.GHz
+
+        assert_quantity_allclose(sq1.doppler_rest, 5 * u.GHz)
+
+        with pytest.raises(ValueError, match="doppler_rest has already been set, and cannot be changed"):
+            sq1.doppler_rest = 4 * u.GHz
+
+        assert_quantity_allclose(sq1.doppler_rest, 5 * u.GHz)
+
+    def test_apply_relativistic_doppler_shift(self):
+
+        # Frequency
+        sq1 = SpectralQuantity(1 * u.GHz)
+        sq2 = sq1._apply_relativistic_doppler_shift(0.5 * c)
+        assert_quantity_allclose(sq2, np.sqrt(1. / 3.) * u.GHz)
+
+        # Wavelength
+        sq3 = SpectralQuantity(500 * u.nm)
+        sq4 = sq3._apply_relativistic_doppler_shift(0.5 * c)
+        assert_quantity_allclose(sq4, np.sqrt(3) * 500 * u.nm)
+
+        # Energy
+        sq5 = SpectralQuantity(300 * u.eV)
+        sq6 = sq5._apply_relativistic_doppler_shift(0.5 * c)
+        assert_quantity_allclose(sq6, np.sqrt(1. / 3.) * 300 * u.eV)
+
+        # Wavenumber
+        sq7 = SpectralQuantity(0.01 / u.micron)
+        sq8 = sq7._apply_relativistic_doppler_shift(0.5 * c)
+        assert_quantity_allclose(sq8, np.sqrt(1. / 3.) * 0.01 / u.micron)
+
+        # Velocity (doppler_convention='relativistic')
+        sq9 = SpectralQuantity(200 * u.km / u.s, doppler_convention='relativistic', doppler_rest=1 * u.GHz)
+        sq10 = sq9._apply_relativistic_doppler_shift(300 * u.km / u.s)
+        assert_quantity_allclose(sq10, 500 * u.km / u.s)
+        assert sq10.doppler_convention == 'relativistic'
+
+        # Velocity (doppler_convention='optical')
+        sq11 = SpectralQuantity(200 * u.km / u.s, doppler_convention='radio', doppler_rest=1 * u.GHz)
+        sq12 = sq11._apply_relativistic_doppler_shift(300 * u.km / u.s)
+        assert_quantity_allclose(sq12, 499.650341 * u.km / u.s)
+        assert sq12.doppler_convention == 'radio'
+
+        # Velocity (doppler_convention='radio')
+        sq13 = SpectralQuantity(200 * u.km / u.s, doppler_convention='optical', doppler_rest=1 * u.GHz)
+        sq14 = sq13._apply_relativistic_doppler_shift(300 * u.km / u.s)
+        assert_quantity_allclose(sq14, 500.350827 * u.km / u.s)
+        assert sq14.doppler_convention == 'optical'
```

Comment:
(Given previous comment, this maybe shouldn't be here)

Also test adding `velocity=0.999*c` twice.

Reply:
Fixed